opennebula/resource_opennebula_template.go

```diff
@@ -150,6 +159,101 @@ func resourceOpennebulaTemplate() *schema.Resource {
 	}
 }
 
+func FeaturesFields() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"pae": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Physical address extension mode allows 32-bit guests to address more than 4 GB of memory.",
+			ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+				validtypes := []string{"YES", "NO"}
+				value := v.(string)
+
+				if inArray(value, validtypes) < 0 {
+					errors = append(errors, fmt.Errorf("PAE must be one of: %s", k, strings.Join(validtypes, ",")))
+				}
+
+				return
+			},
+		},
+		"acpi": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Useful for power management, for example, with KVM guests it is required for graceful shutdown to work.",
+			ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+				validtypes := []string{"YES", "NO"}
+				value := v.(string)
+
+				if inArray(value, validtypes) < 0 {
+					errors = append(errors, fmt.Errorf("ACPI must be one of: %s", k, strings.Join(validtypes, ",")))
+				}
+
+				return
+			},
+		},
+		"apic": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Enables the advanced programmable IRQ management. Useful for SMP machines.",
+			ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+				validtypes := []string{"YES", "NO"}
+				value := v.(string)
+
+				if inArray(value, validtypes) < 0 {
+					errors = append(errors, fmt.Errorf("APIC must be one of: %s", k, strings.Join(validtypes, ",")))
+				}
+
+				return
+			},
+		},
+		"localtime": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "The guest clock will be synchronized to the host’s configured timezone when booted. Useful for Windows VMs.",
+			ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+				validtypes := []string{"YES", "NO"}
+				value := v.(string)
+
+				if inArray(value, validtypes) < 0 {
+					errors = append(errors, fmt.Errorf("LOCALTIME must be one of: %s", k, strings.Join(validtypes, ",")))
+				}
+
+				return
+			},
+		},
+		"hyperv": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Add hyperv extensions to the VM. The options can be configured in the driver configuration, HYPERV_OPTIONS.",
+		},
+		"guest_agent": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Enables the QEMU Guest Agent communication. This only creates the socket inside the VM, the Guest Agent itself must be installed and started in the VM.",
+			ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+				validtypes := []string{"YES", "NO"}
+				value := v.(string)
+
+				if inArray(value, validtypes) < 0 {
+					errors = append(errors, fmt.Errorf("GUEST_AGENT must be one of: %s", k, strings.Join(validtypes, ",")))
+				}
+
+				return
+			},
+		},
+		"virtio_scsi_queues": {
```

Comment:
`virtio_scsi_queues` and `iothreads` should be `TypeInt` I guess but don't know how to skip them in the `getTemplateController` function then if they are not set.

Reply:
I don't have any problem with this
The provider won't check if the user set a number but you could add a method to validate the field instead
In this method you could use a strconv.ParseInt method and just check if it returns an error

Reply:
> I don't have any problem with this The provider won't check if the user set a number but you could add a method to validate the field instead In this method you could use a strconv.ParseInt method and just check if it returns an error

Then I'll keep it like this for now.